core/src/main/scala/com/softwaremill/react/kafka2/Producer.scala

```diff
@@ -34,32 +34,64 @@ class ProducerSendFlowStage[K, V](producerProvider: () => KafkaProducer[K, V])
     val closeTimeout = 60000L
     val producer = producerProvider()
     val logic = new GraphStageLogic(shape) {
+      var awaitingConfirmation = 0L
+      var completionState: Option[Try[Unit]] = None
+
+      def checkForCompletion() = {
+        if (isClosed(messages) && awaitingConfirmation == 0) {
+          completionState match {
+            case Some(Success(_)) => completeStage()
+            case Some(Failure(ex)) => failStage(ex)
+            case None => failStage(new IllegalStateException("Stage completed, but there is no info about status"))
+          }
+        }
+      }
+
+      val decrementConfirmation = getAsyncCallback[Unit] { _ =>
+        awaitingConfirmation -= 1
+        checkForCompletion()
+        ()
+      }
+
       setHandler(confirmation, new OutHandler {
-        override def onPull(): Unit = {
+        override def onPull() = {
           tryPull(messages)
         }
       })
 
       setHandler(messages, new InHandler {
-        override def onPush(): Unit = {
+        override def onPush() = {
           val msg = grab(messages)
           val result = Promise[(ProducerRecord[K, V], RecordMetadata)]
           producer.send(msg, new Callback {
-            override def onCompletion(metadata: RecordMetadata, exception: Exception): Unit = {
+            override def onCompletion(metadata: RecordMetadata, exception: Exception) = {
               val completion = Option(metadata).map(m => Success((msg, m))).getOrElse(Failure(exception))
               result.complete(completion)
+              decrementConfirmation.invoke(())
```

Comment:
In `Sender` there is method `completeBatch` which call `batch.done` which completes Future and call `Callback`